### ISSUE TITLE
Building Conditional Wavenets

### ISSUE THREAD

**Issue by neale** [ibab/tensorflow-wavenet#164]
Has anyone done any preliminary work on conditioning a Wavenet on some other information. If we believe that a Wavenet creates a real conditional distribution (unlike a GAN) then we can use smaller Wavenets as a kind of pool operation that doesn't lose any information. 

My understanding of a context stack is that they learn small portions of the input and tag them as "conditional on x" where x is the portion of the input that the context stack saw. The motivation being that many small Wavenets are more tractable than one large one, especially if information is not lost.  

We just need some mechanism for either parsing the input and tagging sections of the input that will be used to spawn conditional Wavenets, or we should allow a user to tag the input with a string

Thoughts? This is the road I'm planning to embark on and I'd appreciate input.

**Comment by kaihuchen** [comment]
Sounds like a great idea! A single WaveNet will forget what it has learned as the training data shifts and changes, so training a whole set of them letting each one being the expert on only one tag (e.g., tag for the voice of a specific person), let it solidify (i.e., stop learning) after loss is low, then allocate a new WaveNet module when a new tag comes along. To start with, a naive manager for such a set of WaveNet modules could be doing just simple argmax-type operation, but of course it can also be quite sophisticated.

We should try to settle on a suitable test domain first, which will help greatly in grounding the design of such a system.

**Comment by kaihuchen** [comment]
This could work mainly because WaveNet does **generative modeling** which won't over-generalize like **discriminative modeling** would, so it is possible to train one tag, set it aside when done to train another, then eventually put a bunch together to do something useful.

**Comment by jyegerlehner** [comment]
See [issue 112](https://github.com/ibab/tensorflow-wavenet/issues/112).

I don't quite follow your "tagging" description of context stacks.  My vision of a context stack is that it involves two wavenets (though could be generalized to N), call one the "shorter range" wavenet, and the other the "longer range" wavenet. The two wavenets jointly comprise the "context stack":
1) The shorter range wavenet has a receptive field of, say, 300 msec.
2) It locally conditions on an input.
3) That input is the output of the "longer-range" wavenet, which has a receptive field of, say, 5 seconds. So this longer-range wavenet has its output not a softmax, but rather a dense vector. 
4) The longer-range wavenet is smaller channel-wise (has fewer dilation channels), and may involve pooling, or may have a slower sample rate. These differences allow the combined shorter- and longer- range wavenets (the context stack) to be smaller than just a single wavenet whose receptive field would be as large as the longer-range field wavenet.
5) The frequency of the output of the longer-range wavenet may be lower than the shorter-range (due to pooling). Thus there is a learnt upsampling deconvolution (or transpose conv) to match its outputs to the shorter-range in providing the vector upon which the shorter-range wavenet locally conditions.

**Comment by neale** [comment]
@jyegerlehner I agree with you on what a context stack would look like. I'm pretty fuzzy on what N wavenets would look like. Deepmind mentions that they would have varying layers, receptive fields, and sample rates. It seems like a pretty large grid search to find any sort of optimality here. 
With tagging I was just referring to the encoding of speaker identities or music genres as a one hot vector, and globally/locally conditioning on that embedding. But it looks like theres a lot of good work and discussion already in [112](https://github.com/ibab/tensorflow-wavenet/issues/112). Thanks for the heads up.
